Encryption v2

jaschrep-msft opened the pull request:
Adding authenticated encryption support.

## Behavioral Changes
- Ranged Download (`DownloadStreaming`, `DownloadContent`)
  - If client is configured with client-side encryption options, first fetches properties to determine the appropriate range adjustment, then does the ranged read.
  - This does not affect other downloads that use this under the hood
  - Pro: Not excessively extending ranges when unnecessary
  - Con: Extra rest call
- Partitioned Download (`DownloadTo`)
  - Decryption shifted from per-partition to the reassembled content stream
  - Pro: avoids overhead added to ranged download
  - Pro: avoids redownloading overlapping bytes from range extension
  - Pro: effectively separates transfer options from decryption logic
  - Con: v1 decryption loses parallelization (was tied to transfer parallelization)

jaschrep-msft committed:
authenticated encryption infrastructure

Added Stream implementation to buffer an "authenticated region" for
cryptographic transformations, then safely apply the transform on
buffered data before passing on through the stream pipeline

jaschrep-msft committed:
infra for branching encryption path

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Storage.Blobs](https://apiview.dev/Assemblies/Review/94b10c8c369f42a29466ee1e71160902)

jaschrep-msft committed:
fixed a visibility modifier

jaschrep-msft committed:
moved a shared code test to common

jaschrep-msft committed:
placeholder GCM transform class

waiting on GCM exposure to pull into this branch

jaschrep-msft committed:
export api

version enum v2

jaschrep-msft committed:
ClientSideEncryptorV2

Hooked up main functionality for clientside encryptor v2
renamed some values
Made IAuthenticatedCryptoGraphicTransform disposable

jaschrep-msft committed:
decryption branching logic

jaschrep-msft committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into encryption-gcm

jaschrep-msft committed:
blank

jaschrep-msft committed:
add netstandard2.1 target to Azure.Storage.Blobs

jaschrep-msft committed:
initial pass of updating tests

jaschrep-msft committed:
bug fixes and tests

jaschrep-msft committed:
partitioned download v2 working

partitioned download v1 temp disabled

jaschrep-msft committed:
renabled v1 encryption partitioned download

other fixes

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/Azure.Storage.Blobs.csproj:
```diff
@@ -1,12 +1,12 @@
     <Version>12.13.0-beta.1</Version>
     <!--The ApiCompatVersion is managed automatically and should not generally be modified manually.-->
-    <ApiCompatVersion>12.12.0</ApiCompatVersion>
+    <ApiCompatVersion Condition="'$(TargetFramework)' != 'netstandard2.1'">12.12.0</ApiCompatVersion>
```
Temporary workaround from AzureSDK until an initial version with a netstandard2.1 target is on NuGet.

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/TransformMode.cs:
```diff
@@ -0,0 +1,27 @@
+namespace Azure.Storage.Cryptography.Models
+{
+    /// <summary>
+    /// Transformation mode for a give <see cref="IAuthenticatedCryptographicTransform"/>.
```
"given"

jaschrep-msft committed:
fix a hardcoded v1 invokcation

jaschrep-msft committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into encryption-gcm

Merge to get bcrypt access to aes in standard2.0

jaschrep-msft committed:
project and test fixes

blobs and queues now only allow unsafe in netstandard2.0
common.tests now allows unsafe in net461
gcm supported on netstandard2.0
fixed clientside encryption tests on net461

jaschrep-msft committed:
removed completed TODOs

jaschrep-msft commented:
@KrzysztofCwalina @tg-msft PR is in a state where your interests can be reviewed. You'll be interested in:

- csproj files
- Azure.Storage.Common.netstandard2.0.cs
- GcmAuthenticatedCryptographicTransform.cs
- BlobBaseClient.cs

Behavioral changes listed in PR description. Contents of Shared/ClientSideEncryption is where all the heavy operations take place if you need context. Happy to answer any questions.

QueueClient work is upcoming. Since queue message size is trivial that they are already buffered in the first place, there are no surprises coming in how data is managed.

jaschrep-msft committed:
added encryption v2 to queues

queue client now encrypts in specified version
fixed a bug in queue client tests
added v2 tests

jaschrep-msft committed:
test assertion

jaschrep-msft committed:
versioned more blob tests

jaschrep-msft committed:
fixed openread bug

jaschrep-msft committed:
closing brace

jaschrep-msft committed:
fixed tests

jaschrep-msft committed:
encryption json schema change

jaschrep-msft committed:
openwrite encryption tests

fixed bug in AuthenticatedRegionCryptoStream write mode.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1398,12 +1402,43 @@ private void SetNameFieldsIfNull()
             bool rangeGetContentHash,
             IProgress<long> progressHandler,
             string operationName,
+            bool bypassClientSideEncryption,
```
Ya I like that better

amnguye commented on sdk/storage/Azure.Storage.Common/tests/Azure.Storage.Common.Tests.csproj:
```diff
@@ -8,6 +8,14 @@
   </PropertyGroup>
+
+  <PropertyGroup Condition="'$(TargetFramework)' == 'net461'">
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Should we eventually move these tests to the Aes Gcm Tests package to avoid having to do this in the Common Tests package?

jaschrep-msft committed:
open write bug fix

jaschrep-msft committed:
only flush on dispose when in write mode

amnguye commented on sdk/storage/Azure.Storage.Blobs/tests/BlobClientOpenWriteTests.cs:
```diff
@@ -18,7 +18,7 @@ namespace Azure.Storage.Blobs.Tests
     {
         public BlobClientOpenWriteTests(bool async, BlobClientOptions.ServiceVersion serviceVersion)
-            : base(async, serviceVersion, null /* RecordedTestMode.Record /* to re-record */)
+            : base(async, serviceVersion, RecordedTestMode.Record /* RecordedTestMode.Record /* to re-record */)
```
Remember to change back to null

jaschrep-msft committed:
encryption metadata format design change

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/tests/AuthenticatedRegionCryptoStreamTest.cs:
```diff
@@ -0,0 +1,432 @@
+            }
+        }
+
+#if NETCOREAPP3_1
```
Remove conditional compilation.

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/tests/AuthenticatedRegionCryptoStreamTest.cs:
```diff
@@ -0,0 +1,432 @@
+#if NETCOREAPP3_1
+        [Test]
+        [Combinatorial]
+        public void RoundTrip_Read(
+            [Values(true, false)] bool useRealEncryption,
+            [Values(true, false)] bool alligned,
+            [Values(1, 3)] int numAuthBlocks)
```
@amnguye this is the one test that needs the unsafe tag to be added, and only for `useRealEncryption == true`. I'm not sure just removing a test is a good idea, but I'll admit it becomes more of an integration test than a unit test when we use GCM here. Perhaps we need a broader effort to classify tests as unit/integration/smoke. This would likely cut down on bulky test classes and packages in general.

jaschrep-msft replied:
I have moved it to the blob encryption tests, which already need to be configured for something like this. I can take a work item later to revisit our testing organization.

jaschrep-msft committed:
test cleanup

jaschrep-msft committed:
pr comments

kasobol-msft commented on sdk/storage/Azure.Storage.Queues/src/Azure.Storage.Queues.csproj:
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
```
should this get compiled to netstandard2.1?

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1424,11 +1459,14 @@ private void SetNameFieldsIfNull()
 
-                    if (UsingClientSideEncryption)
+                    if (UsingClientSideEncryption && !ignoreClientSideEncryption)
                     {
                         // TODO #27253
                         //options = BlobDownloadOptions.CloneOrDefault(options) ?? new BlobDownloadOptions();
-                        range = BlobClientSideDecryptor.GetEncryptedBlobRange(range);
+                        if ((await GetPropertiesInternal(conditions, async, cancellationToken).ConfigureAwait(false)).Value.Metadata.TryGetValue(Constants.ClientSideEncryption.EncryptionDataKey, out string rawEncryptiondata))
+                        {
+                            range = BlobClientSideDecryptor.GetEncryptedBlobRange(range, rawEncryptiondata);
+                        }
                     }
```
is there other reason than get properties call for `ignoreClientSideEncryption` ? 
It seems that we're pushing up implementation to callers where properties are already available.

Would passing optional properties/metadata to this method eliminate need for  `ignoreClientSideEncryption` flag? I.e. call getProperties if needed?

jaschrep-msft replied:
It ignores overlapping ranges for something like a partitioned download, and also just skips decryption as a whole. This was trivial in v1 when ranges were extended a handful of bytes, but now we are extending to ~4MB bounds, which may not play nice with transfer options. While we could pass in optional prefetched properties, we also ignore potential boundary extension waste by just shifting decryption back to the partitioned downloader.

jaschrep-msft replied:
I do see how just adding bool flags is unfun, though. Perhaps it's better for the "real" DownloadStreamingInternal method as it exists today to not have a concept of client-side encryption. Instead, the public ranged download APIs go through a "DirectCalledDownloadStreamingInternal" which applies the appropriate transformations. All customer-called ranged downloads go through there, and our partitioned downloader and open-read don't. This way every public API goes through some function that applies client-side encryption, but we don't have flagged branching logic in our low-level REST download. 

Thoughts?

kasobol-msft replied:
yeah, that'd be my next suggestion. `UsingClientSideEncryption` and `ignoreClientSideEncryption` in same condition looks odd and smells like spaghetti.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/tests/PartitionedDownloaderTests.cs:
```diff
@@ -205,32 +201,60 @@ private void AssertContent(int expectedLength, MemoryStream stream)
-            }
+                });
+
+            //if (_async)
+            //{
+            //    blockClient.Setup(c => c.DownloadStreamingAsync(It.IsAny<HttpRange>(), It.IsAny<BlobRequestConditions>(), false, s_cancellationToken))
+            //        .Returns<HttpRange, BlobRequestConditions, bool, CancellationToken>((range, conditions, rangeGetHash, cancellation) =>
+            //            dataSource.GetStreamAsync(
+            //                range,
+            //                conditions,
+            //                rangeGetHash,
+            //                progress: default,
+            //                cancellation));
+            //    blockClient.Setup(c => c.DownloadStreamingAsync(It.IsAny<HttpRange>(), It.IsAny<BlobRequestConditions>(), false, It.IsAny<IProgress<long>>(), s_cancellationToken))
+            //        .Returns<HttpRange, BlobRequestConditions, bool, IProgress<long>, CancellationToken>(dataSource.GetStreamAsync);
+            //}
+            //else
+            //{
+            //    blockClient.Setup(c => c.DownloadStreaming(It.IsAny<HttpRange>(), It.IsAny<BlobRequestConditions>(), false, s_cancellationToken))
+            //        .Returns<HttpRange, BlobRequestConditions, bool, CancellationToken>((range, conditions, rangeGetHash, cancellation) =>
+            //            dataSource.GetStream(
+            //                range,
+            //                conditions,
+            //                rangeGetHash,
+            //                progress: default,
+            //                cancellation));
+            //    blockClient.Setup(c => c.DownloadStreaming(It.IsAny<HttpRange>(), It.IsAny<BlobRequestConditions>(), false, It.IsAny<IProgress<long>>(), s_cancellationToken))
+            //        .Returns<HttpRange, BlobRequestConditions, bool, IProgress<long>, CancellationToken>(dataSource.GetStream);
+            //}
```
Remove

jaschrep-msft committed:
fixes/recordings

jaschrep-msft committed:
rerecord

jaschrep-msft committed:
moved test

jaschrep-msft committed:
PR refactor comments

jaschrep-msft committed:
fixed mock setup

Pablo1q reviewed (approved):
Good

jaschrep-msft commented:
/azp run net - storage - tests

jaschrep-msft committed:
macos test pipeline installs OpenSSL

jaschrep-msft commented on sdk/storage/tests.yml:
```diff
@@ -6,13 +6,13 @@ extends:
-    Location: canadacentral
+    Location: canadacentral
     CloudConfig:
       Preview:
-        SubscriptionConfiguration: $(sub-config-azure-cloud-test-resources-preview)
-      PrivatePreview:
+        SubscriptionConfiguration: $(sub-config-azure-cloud-test-resources-preview)
+      PrivatePreview:
         SubscriptionConfiguration: $(sub-config-storage-test-resources)
-    Clouds: Preview
```
line ending normalization from editor

weshaggard commented on sdk/storage/tests.yml:
```diff
@@ -26,5 +26,9 @@ extends:
         AZURE_ONLY_TEST_LATEST_SERVICE_VERSION: true
+    PreSteps:
+      - bash: brew install openssl@1.1
+        displayName: (MacOS) Grant access to ~/.Azure
```
Probably should update to display name. 
```suggestion
        displayName: (MacOS) Install OpenSSL
```

jaschrep-msft committed:
Update sdk/storage/tests.yml

Co-authored-by: Wes Haggard <weshaggard@users.noreply.github.com>

jaschrep-msft committed:
bind content encryption algorithm to wrapped key

jaschrep-msft committed:
tests.yml update

now installs openssl3
updates an envvar for .net to find the openssl implementation

jaschrep-msft committed:
Merge branch 'encryption-gcm' of https://github.com/jaschrep-msft/azure-sdk-for-net into encryption-gcm

jaschrep-msft committed:
added "tests" steps to "ci"

some fixes

jaschrep-msft commented:
/azp run net - storage - ci

jaschrep-msft committed:
indentation error

jaschrep-msft committed:
trying different conditional syntax

amnguye commented:
/azp run net - storage - ci

weshaggard commented on sdk/storage/ci.yml:
```diff
@@ -53,3 +55,5 @@ extends:
       safeName: AzureResourceManagerStorage
     TestSetupSteps:
     - template: /sdk/storage/tests-install-azurite.yml
+    - ${{ contains(variables['OSVmImage'], 'mac') }}:
+        bash: brew install openssl@3
```
```suggestion
    - bash: brew install openssl@3
      displayName: (MacOS) Install OpenSSL
      condition: contains(variables['OSVmImage'], 'mac')
    
```

weshaggard committed:
Fix syntax for openssl install step

jaschrep-msft committed:
yaml changes

reverted ci.yml
rearranged tests.yml

jaschrep-msft committed:
Merge branch 'encryption-gcm' of https://github.com/jaschrep-msft/azure-sdk-for-net into encryption-gcm

jaschrep-msft committed:
set a test to live only to play nice with pipeline

jaschrep-msft committed:
ignore a half-ready test

jaschrep-msft commented:
/azp run net - storage - tests

tg-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/UriExtensions.cs:
```diff
@@ -113,7 +113,7 @@ public static string GetAccountNameFromDomain(string host, string serviceSubDoma
         public static bool IsHostIPEndPointStyle(this Uri uri) =>
             (!string.IsNullOrEmpty(uri.Host) &&
-            uri.Host.IndexOf(".", StringComparison.InvariantCulture) >= 0 &&
+            uri.Host.Contains(".") &&
```
Why was this change necessary?  Can we pass the invariant culture into `Contains` at least?

jaschrep-msft replied:
Netstandard2.1 does not like `IndexOf()` and warns. I can try and put invariant back in there.

tg-msft commented on sdk/storage/Azure.Storage.Common/src/ClientSideEncryptionVersion.cs:
```diff
@@ -12,7 +12,12 @@ public enum ClientSideEncryptionVersion
         /// 1.0
         /// </summary>
-        V1_0 = 1
+        V1_0 = 1,
```
Should we add a comment about which version to prefer?

tg-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/AuthenticatedRegionCryptoStream.cs:
```diff
@@ -0,0 +1,290 @@
+        // need to record how much data in buffer is legitimate
+        private int _bufferPopulatedLength;
+
+        private readonly int _tempRefilBufferSize;
```
Nit - Refill typo?

tg-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/ClientSideEncryptorV2_0.cs:
```diff
@@ -0,0 +1,199 @@
+
+        private void ValidateMembers()
+        {
+            if (_keyEncryptionKey == default || _keyWrapAlgorithm == default)
+            {
+                throw Errors.ClientSideEncryption.MissingRequiredEncryptionResources(nameof(_keyEncryptionKey), nameof(_keyWrapAlgorithm));
+            }
```
Can we do this validation just once in the .ctor?

tg-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/IAuthenticatedCryptographicTransform.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Azure.Storage.Cryptography
+{
+    internal interface IAuthenticatedCryptographicTransform : IDisposable
```
Could we get away with making these `abstract class`es instead?  It probably doesn't matter as much for shared source, but versioning is always more flexible with an `abstract class`.

jaschrep-msft replied:
I'm actually a bit afraid of flexibility here for this shared source. This is delicate code and I wouldn't want any assumptions to be made when adding a new piece of functionality.

tg-msft reviewed (approved):
Looks great!  I left a few nits, but nothing critical that needs to be addressed immediately.

jaschrep-msft committed:
removed unnecessary test

jaschrep-msft committed:
Merge branch 'encryption-gcm-with-algorithmwrapping' into encryption-gcm

jaschrep-msft committed:
tests

jaschrep-msft commented:
@tg-msft Good catches! All internals though that will need to wait for GA. I am already writing my own list of work to track and I will add these to the list. Issue will be posted before merge.

jaschrep-msft committed:
record

jaschrep-msft committed:
changelog

jaschrep-msft committed:
Merge branch 'main' into encryption-gcm

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/AuthenticatedRegionCryptoStream.cs:
```diff
@@ -0,0 +1,290 @@
+                throw Errors.InvalidArgument(nameof(transform));
+            }
+
+            _buffer = new byte[bufferSize];
```
This particular buffer is long-living (life of the stream) and we need it to be exact size (not guaranteed by array pool) or the math gets even more complex. But there are several other temporary arrays throughout this class that can be taken from the array pool.
